File: cpr-fantasy-football/lib/data-fetcher.ts
import Papa from 'papaparse';

// Simple in-memory cache with expiration
interface CacheEntry<T> {
  data: T[];
  timestamp: number;
}

const cache = new Map<string, CacheEntry<unknown>>();
const CACHE_DURATION_MS = 60000; // 1 minute cache

/**
 * Generic CSV fetcher using PapaParse with in-memory caching
 * @param url - The URL of the CSV file to fetch
 * @param useHeaders - Whether the CSV has headers (default: true). If false, columns are accessed by index
 * @returns Promise resolving to array of parsed CSV rows
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function fetchCSV<T = any>(url: string, useHeaders: boolean = true): Promise<T[]> {
  const MAX_RETRIES = 3;
  const RETRY_DELAY = 1000; // 1 second
  const TIMEOUT = 10000; // 10 seconds

  let lastError: Error | null = null;

  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    try {
      // Check cache first
      const cacheKey = `${url}-${useHeaders}`;
      const cached = cache.get(cacheKey) as CacheEntry<T> | undefined;

      if (cached && Date.now() - cached.timestamp < CACHE_DURATION_MS) {
        console.log(`Using cached data for ${url}`);
        return cached.data;
      }

      // Fetch fresh data
      console.log(`Fetching fresh data for ${url} (Attempt ${attempt}/${MAX_RETRIES})`);

      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), TIMEOUT);

      try {
        const response = await fetch(url, {
          cache: 'no-store',
          signal: controller.signal,
        });

        if (!response.ok) {
          throw new Error(`Failed to fetch CSV: ${response.statusText} (${response.status})`);
        }

        const csvText = await response.text();

        const data = await new Promise<T[]>((resolve, reject) => {
          Papa.parse(csvText, {
            header: useHeaders,
            dynamicTyping: true,
            skipEmptyLines: true,
            complete: (results) => {
              if (results.errors.length > 0) {
                console.warn(`CSV Parsing warnings for ${url}:`, results.errors);
              }
              resolve(results.data as T[]);
            },
            error: (error: Error) => {
              reject(error);
            },
          });
        });

        // Store in cache
        cache.set(cacheKey, {
          data,
          timestamp: Date.now(),
        });

        return data;
      } finally {
        clearTimeout(timeoutId);
      }
    } catch (error) {
      console.error(`Error fetching CSV (Attempt ${attempt}/${MAX_RETRIES}):`, error);
      lastError = error instanceof Error ? error : new Error(String(error));

      if (attempt < MAX_RETRIES) {
        await new Promise(resolve => setTimeout(resolve, RETRY_DELAY * attempt));
      }
    }
  }

  throw lastError || new Error('Failed to fetch CSV after multiple attempts');
}

/**
 * Clean and normalize strings for comparison
 * @param str - String to normalize
 * @returns Trimmed lowercase string
 */
export function normalizeString(str: string | null | undefined): string {
  if (!str) return '';
  return str.trim().toLowerCase();
}

/**
 * Parse numeric values from various formats
 * Handles currency symbols (£, $) and comma separators
 * @param value - Value to parse (string or number)
 * @returns Parsed number, or 0 if parsing fails
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function parseNumber(value: any): number {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    // Remove currency symbols and commas
    const cleaned = value.replace(/[£$,]/g, '');
    const parsed = parseFloat(cleaned);
    return isNaN(parsed) ? 0 : parsed;
  }
  return 0;
}

File: conda-fantasy-football/src/data_fetcher.py
"""Data fetching utilities for CPR Fantasy Football."""

import pandas as pd
import streamlit as st
from datetime import datetime, timedelta
from typing import Optional
import re


@st.cache_data(ttl=60)  # Cache for 60 seconds
def fetch_csv(url: str, skip_rows: int = 0) -> pd.DataFrame:
    """
    Fetch and parse CSV data from a URL with caching.

    Args:
        url: The URL of the CSV file to fetch
        skip_rows: Number of rows to skip at the beginning (default: 0)

    Returns:
        DataFrame containing the parsed CSV data
    """
    try:
        df = pd.read_csv(url, skiprows=skip_rows)
        return df
    except Exception as e:
        st.error(f"Error fetching CSV from {url}: {str(e)}")
        raise


def normalize_string(value: Optional[str]) -> str:
    """
    Clean and normalize strings for comparison.

    Args:
        value: String to normalize

    Returns:
        Trimmed lowercase string
    """
    if value is None or pd.isna(value):
        return ""
    return str(value).strip().lower()


def parse_number(value) -> float:
    """
    Parse numeric values from various formats.
    Handles currency symbols (£, $) and comma separators.

    Args:
        value: Value to parse (string or number)

    Returns:
        Parsed number, or 0.0 if parsing fails
    """
    if pd.isna(value):
        return 0.0

    if isinstance(value, (int, float)):
        return float(value)

    if isinstance(value, str):
        # Remove currency symbols and commas
        cleaned = re.sub(r'[£$,]', '', value)
        try:
            return float(cleaned)
        except ValueError:
            return 0.0

    return 0.0


def clean_player_name(name) -> str:
    """
    Clean and normalize player names.

    Args:
        name: Player name to clean

    Returns:
        Cleaned player name as string
    """
    if pd.isna(name):
        return ""
    return str(name).strip()
